### PR TITLE
Rust telemetry to JSON files, instead of OTLP/gRPC

### PR DESCRIPTION
**Pull request by graebm**
Lots going on here, apologies...

- Write telemetry to file, instead of exporting via OTLP/gRPC via [opentelemetry-otlp](https://crates.io/crates/opentelemetry-otlp/) crate.
  - Reasons for change:
    - Exporting via OTLP/gRPC was impacting noticeably performance. I tried to improve this by delaying any actual work until the benchmark run was complete (increasing buffer sizes, manually flushing buffers), but it still slowed max throughput from ~70 -> ~50 Gb/s.
    - It's just much simpler to spit out a file
  - Add our own telemetry exporter code, that writes [OpenTelemetry Protocol](https://opentelemetry.io/docs/specs/otel/protocol/) (OTLP) in JSON format.
      - We started with code from an old version of the `opentelemetry-stdout` crate. Specifically, we took the commit just before the [Simplify LogExporter::Export interface](https://github.com/open-telemetry/opentelemetry-rust/commit/3193320fa6dc17e89a7bed6090000aef781ac29c) change. This change removed JSON export, so the code would be simpler for others to copy/paste. But we wanted JSON export, so copy/pasting from that point.
      - Modify the exporter so it doesn't do ANY work (buffering up all spans) until `flush_to_file(filename)` is called.
        - Hopefully, this eliminates the performance impact of gathering telemetry
        - This gives us a unique file per run. It's much nicer to analyze a run on its own, vs trying to isolate run 5 of 10 within an enormous file.
        - File is named like `trace_20241009T185957Z_download-30GiB-1x-ram_run01.json`
- Add python scripts to visualize the telemetry data
    - Start with `allspans.py` which graphs ALL tracing spans, each in its own row, similar to tools like [Jaeger](https://github.com/jaegertracing/jaeger-ui)
        - Draw with [`plotly.express.timeline()`](https://plotly.com/python/gantt/).
            - `plotly` was recommended by ChatGPT, I asked for a python graphing library where you could hide information until you mouse-over it. I also specified that files should be easy to share, and `plotly` generates a single HTML file.
        - The figure could use more work, but it's a start
    - More visualizations coming...

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by graebm** [comment]
Screenshot of small workload: 24MiB (3 parts) download:
Screenshot 2024-10-14 at 11 44 38 AM

Screenshot of large workload: 30GiB (3840 parts) download:
Screenshot 2024-10-14 at 11 44 17 AM

File: runners/s3-benchrunner-rust/src/telemetry/trace/exporter.rs [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+impl opentelemetry_sdk::export::trace::SpanExporter for SpanExporter {
+    fn export(&mut self, batch: Vec<export::trace::SpanData>) -> BoxFuture<'static, ExportResult> {
+        // Queue batch, along with the current resource
+        let batch = SdkSpanDataBatch {
+            resource: self.resource.clone(),
+            batch: batch,
+        };
+        self.queued_batches.lock().unwrap().push(batch);
```
Comment (waahm7) [inline runners/s3-benchrunner-rust/src/telemetry/trace/exporter.rs:42]: Potential optimization todo: we can pre-allocate the vector to a big number.

**Review by waahm7 — APPROVED** [review]
image

File: runners/s3-benchrunner-rust/graph/allspans.py [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+
+    # prepare columns for plotly
+    columns = defaultdict(list)
+    name_count = defaultdict(int)
+    for (idx, span) in enumerate(spans):
+
+        name = span['name']
+        # we want each span in its own row, so assign a unique name and use that as Y value
+        name_count[name] += 1
+        unique_name = f"{name}#{name_count[name]}"
```
[the remark this answers is not in the document]
Comment (graebm) [reply runners/s3-benchrunner-rust/graph/allspans.py:28]: added TODO

**Review by aajtodd — APPROVED** [review]

**Pull request merged**